File: Jump_Game_II.cpp
//Time Complexity-O(n)
//Space Complexity-O(1)
//Did the code execute on Leetcode? Yes

class Solution {
public:
    int jump(vector<int>& nums) {
        if(nums.size()==0|| nums.size()==1)
        {
            return 0;
        }
        int current=nums[0];
        int next=nums[0];
        int count=1;
        for(int i=1;i<nums.size();i++)
        {
            next=max(next,i+nums[i]);
            if(i<nums.size()-1 && i==current)
            {
                current=next;
                count++;
            }
        }
        return count;
    }
};

File: jumpGameii.java
class Solution {
    public int jump(int[] nums) {
        // Tc: O(n) Sc: O(1)
        int n = nums.length;
        int cnt = 0;
        int dest = 0;
        int window = 0;

        for (int i = 0; i < n - 1; i++) {
            window = Math.max(window, i + nums[i]);

            if (i == dest) {
                cnt++;
                dest = window;
            }
        }
        return cnt;
    }
}